### PR TITLE
Adding initial radiona ULX4M-LD-V2

### PR DESCRIPTION
**Pull request by goran-mahovlic**

### CODE REVIEW

**Comment by enjoy-digital** [comment]
Hi @goran-mahovlic,

thanks a lot. The platfom file seems to be missing, can you add it to the PR?

**Comment by goran-mahovlic** [comment]
> Hi @goran-mahovlic,
> 
> thanks a lot. The platfom file seems to be missing, can you add it to the PR?

Huh :) I was doing something else so did few extra commits. Hope it is good now

File: litex_boards/targets/ulx4m_ld_v2.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+        # GPIOs ------------------------------------------------------------------------------------
+        if with_syzygy_gpio:
+            platform.add_extension(ulx4m_ld_v2.raw_syzygy_io("SYZYGY0"))
+            self.submodules.gpio = GPIOTristate(platform.request("SYZYGY0"))
+
+# Build --------------------------------------------------------------------------------------------
+
+def main():
+    parser = argparse.ArgumentParser(description="LiteX SoC on ULX4M-LD-V2")
```
Comment (trabucayre) [inline litex_boards/targets/ulx4m_ld_v2.py:195]: Small remark: why not using `LiteXArgumentParser` as done for [radiona_ulx3s](https://github.com/litex-hub/litex-boards/blob/master/litex_boards/targets/radiona_ulx3s.py#L149).
This reduce (a bit) the list arguments to be explicitly added.

Comment (goran-mahovlic) [reply litex_boards/targets/ulx4m_ld_v2.py:195]: Yes I also need to rename file - Sorry for problems I am doing editing directly as I have more not important changes locally, like v1 and SDRAM version. Will change as you suggested...

Comment (goran-mahovlic) [reply litex_boards/targets/ulx4m_ld_v2.py]: @trabucayre do you thing it is ok now?

File: litex_boards/platforms/radiona_ulx4m_ld_v2.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+class Platform(LatticePlatform):
+    default_clk_name   = "clk25"
+    default_clk_period = 1e9/25e6
+
+    def __init__(self, revision="0.1", device="85F", **kwargs):
+        assert revision in ["0.1"]
+        self.revision = revision
+        io         = {"0.1": _io_r0_1}[revision]
+        connectors = {"0.1": _connectors_r0_1}[revision]
+        LatticePlatform.__init__(self, f"LFE5UM5G-85F-8BG381C", io, connectors, **kwargs)
```
Comment (trabucayre) [inline litex_boards/platforms/radiona_ulx4m_ld_v2.py:170]: ditto

File: litex_boards/platforms/radiona_ulx4m_ld_v2.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+#    def create_programmer(self):
+#        return DFUProg(vid="1209", pid="5af0")
+
+    def do_finalize(self, fragment):
+        LatticePlatform.do_finalize(self, fragment)
```
Comment (trabucayre) [inline litex_boards/platforms/radiona_ulx4m_ld_v2.py:176]: ditto

File: litex_boards/targets/radiona_ulx4m_ld_v2.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+# Copyright (c) 2021 Greg Davill <greg.davill@gmail.com>
+# Copyright (c) 2022 Goran Mahovlic <goran.mahovlic@gmail.com>
+# SPDX-License-Identifier: BSD-2-Clause
+
+# Build/Use:
+# ./gsd_ulx4m_ld.py  --uart-name=uart --uart-baudrate=115200 --sdram-device MT41K64M16 --csr-csv=csr.csv --build
+
+import os
+import sys
+import argparse
```
Comment (trabucayre) [inline litex_boards/targets/radiona_ulx4m_ld_v2.py:16]: no more required

File: litex_boards/targets/radiona_ulx4m_ld_v2.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            platform.add_extension(ulx4m_ld_v2.raw_syzygy_io("SYZYGY0"))
+            self.submodules.gpio = GPIOTristate(platform.request("SYZYGY0"))
+
+# Build --------------------------------------------------------------------------------------------
+
+def main():
+    parser = LiteXArgumentParser(platform=radiona_ulx4m_ld_v2.Platform, description="LiteX SoC on ULX4M-LD-V2")    
+    parser.add_argument("--build",           action="store_true",    help="Build bitstream.")
+    parser.add_argument("--load",            action="store_true",    help="Load bitstream.")
+    parser.add_argument("--toolchain",       default="trellis",      help="FPGA toolchain (trellis or diamond).")
```
Comment (trabucayre) [inline litex_boards/targets/radiona_ulx4m_ld_v2.py:198]: These options are automatically injected by LitexArgumentParser

File: litex_boards/targets/radiona_ulx4m_ld_v2.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    sdopts = parser.add_mutually_exclusive_group()
+    sdopts.add_argument("--with-spi-sdcard", action="store_true", help="Enable SPI-mode SDCard support.")
+    sdopts.add_argument("--with-sdcard",     action="store_true", help="Enable SDCard support.")
+    parser.add_argument("--with-syzygy-gpio",action="store_true", help="Enable GPIOs through SYZYGY Breakout on Port-A.")
+    viopts = parser.add_mutually_exclusive_group()
+    viopts.add_argument("--with-video-terminal",    action="store_true", help="Enable Video Terminal (HDMI).")
+    viopts.add_argument("--with-video-framebuffer", action="store_true", help="Enable Video Framebuffer (HDMI).")
+    builder_args(parser)
+    soc_core_args(parser)
+    trellis_args(parser)
```
Comment (trabucayre) [inline litex_boards/targets/radiona_ulx4m_ld_v2.py:218]: automatically done by `LitexArgumentParser` too

File: litex_boards/targets/radiona_ulx4m_ld_v2.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        with_spi_flash   = args.with_spi_flash,
+        with_video_terminal    = args.with_video_terminal,
+        with_video_framebuffer = args.with_video_framebuffer,
+        with_syzygy_gpio = args.with_syzygy_gpio,
+        **soc_core_argdict(args))
+    if args.with_spi_sdcard:
+        soc.add_spi_sdcard()
+    if args.with_sdcard:
+        soc.add_sdcard()
+    builder = Builder(soc, **builder_argdict(args))
```
Comment (trabucayre) [inline litex_boards/targets/radiona_ulx4m_ld_v2.py:242]: s/`**builder_argdict(args)`/ **parser.builder_argdict/g

File: litex_boards/targets/radiona_ulx4m_ld_v2.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    builder_kargs = trellis_argdict(args) if args.toolchain == "trellis" else {}
+    builder.build(**builder_kargs, run=args.build)
```
Comment (trabucayre) [inline litex_boards/targets/radiona_ulx4m_ld_v2.py:244]:
```
if args.build:
     builder.build(**parser.toolchain_argdict)
```

**Comment by enjoy-digital** [comment]
Thanks @goran-mahovlic and @trabucayre, I'm merging to look at @goran-mahovlic's issues with Linux-on-LiteX-Vexrisc. @trabucayre: I'll do the changes you suggested myself :)

**Pull request merged**